packages/core/src/editor/NodeHelpers.ts

```diff
@@ -0,0 +1,189 @@
+import { EditorState, Node, NodeId } from "@craftjs/core";
+import invariant from "tiny-invariant";
+import {
+  deprecationWarning,
+  ERROR_CANNOT_DRAG,
+  ERROR_DUPLICATE_NODEID,
+  ERROR_INVALID_NODE_ID,
+  ERROR_MOVE_INCOMING_PARENT,
+  ERROR_MOVE_NONCANVAS_CHILD,
+  ERROR_MOVE_OUTGOING_PARENT,
+  ERROR_MOVE_TO_DESCENDANT,
+  ERROR_MOVE_TO_NONCANVAS_PARENT,
+  ERROR_MOVE_TOP_LEVEL_NODE,
+  ROOT_NODE,
+} from "@craftjs/utils";
+import { serializeNode } from "../utils/serializeNode";
+import { mergeTrees } from "../utils/mergeTrees";
+
+export function NodeHelpers(state: EditorState, id: NodeId) {
+  invariant(typeof id == "string", ERROR_INVALID_NODE_ID);
+
+  const node = state.nodes[id];
+
+  const ref = (id) => NodeHelpers(state, id);
+
+  const getNodeFromIdOrNode = (node: NodeId | Node) =>
+    typeof node === "string" ? state.nodes[node] : node;
+
+  return {
+    isCanvas() {
+      return !!node.data.isCanvas;
+    },
+    isRoot() {
+      return node.id === ROOT_NODE;
+    },
+    isLinkedNode() {
+      return (
+        node.data.parent &&
+        ref(node.data.parent).linkedNodes().includes(node.id)
+      );
+    },
+    isTopLevelNode() {
+      return !node.data.parent;
+    },
+    isDeletable() {
+      return !this.isTopLevelNode();
+    },
+    isParentOfTopLevelNodes: () => !!node.data.linkedNodes,
+    isParentOfTopLevelCanvas() {
+      deprecationWarning("query.node(id).isParentOfTopLevelCanvas", {
+        suggest: "query.node(id).isParentOfTopLevelNodes",
+      });
+      return this.isParentOfTopLevelNodes();
+    },
+    get() {
+      return node;
+    },
+    ancestors(deep = false) {
+      function recursive(id: NodeId, result: NodeId[] = [], depth: number = 0) {
+        result.push(id);
+        const node = state.nodes[id];
+        if (deep || (!deep && depth === 0)) {
+          if (node.data.parent) {
+            result = recursive(node.data.parent, result, depth + 1);
+          }
+        }
+        return result;
+      }
+      return recursive(node.data.parent);
+    },
+    descendants(deep = false, includeLinkedNodes = false) {
+      function recursive(id: NodeId, result: NodeId[] = [], depth: number = 0) {
+        const node = state.nodes[id];
+        if (deep || (!deep && depth === 0)) {
+          if (ref(id).linkedNodes().length > 0) {
+            ref(id)
+              .linkedNodes()
+              .forEach((nodeId) => {
+                result.push(nodeId);
+                result = recursive(nodeId, result, depth + 1);
+              });
+          }
+
+          if (node.data.nodes) {
+            const childNodes = node.data.nodes; // ['t1c1', 't1c2']
+            childNodes.forEach((nodeId) => {
+              result.push(nodeId);
+              result = recursive(nodeId, result, depth + 1);
+            });
+          }
+        }
+        return result;
+      }
+      return recursive(id);
+    },
+    linkedNodes() {
```

Comment:
hmm what are linked nodes? I don't think I've seen this before

Reply:
Please check the updated [docs](https://deploy-preview-74--craftjs.netlify.app/r/docs/concepts/user-components#defining-editable-elements. It's basically when you define Nodes (using the `<Element />`)  inside a User Component's render, they are different from child Nodes in that they are not passed to the component's render.